mirrord/agent/src/watched_task.rs

```diff
@@ -0,0 +1,102 @@
+use std::{
+    future::Future,
+    sync::{Arc, RwLock, RwLockReadGuard},
+};
+
+use crate::error::AgentError;
+
+/// A shared clonable view on a background task's status.
+#[derive(Debug, Clone)]
+pub struct TaskStatus {
+    /// Name of the task.
+    task_name: &'static str,
+    /// An error that occurred during the execution of the task.
+    error: Arc<RwLock<Option<AgentError>>>,
+}
+
+impl TaskStatus {
+    /// Check if an error occurred during the execution of the task.
+    /// If yes, return [`AgentError::BackgroundTaskFailed`].
+    pub fn check(&self) -> Option<AgentError> {
+        self.error
+            .read()
+            .expect("task status lock poisoned")
+            .as_ref()
+            .map(|e| AgentError::BackgroundTaskFailed {
+                task: self.task_name,
+                cause: e.to_string(),
+            })
+    }
```

Comment:
Another thought: In all the cases where we call `check` we actually know that the task is either finished or is about to (since its side of a channel was dropped), right?
So isn't there a way to have `WatchedTask::start` set the status when it's done in a way where `check` can wait for the status to be set by `start`? That way we could eliminate the `Option`s and defaults altogether...

Reply:
I changed the `TestStatus` interface so that its methods wait for the task to complete before returning the error. Turns out that `tokio` is in dramatic need of some simple condvar primitive, I had to use `tokio::sync::Watch`.